src/connector/src/sink/elasticsearch_opensearch_common.rs

```diff
@@ -0,0 +1,253 @@
+    #[serde(rename = "index_column")]
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    pub index_column: Option<usize>,
+
+    #[serde(rename = "max_task_num")]
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    pub max_task_num: Option<usize>,
+}
+
+impl ElasticSearchOpenSearchConfig {
+    pub fn from_btreemap(
+        mut properties: BTreeMap<String, String>,
+        schema: &Schema,
+    ) -> Result<Self> {
+        let index_column = properties
+                .get(ES_OPTION_INDEX_COLUMN)
+                .cloned()
+                .map(|n| {
+                    schema
+                        .fields()
+                        .iter()
+                        .position(|s| s.name == n)
+                        .ok_or_else(|| anyhow!("please ensure that '{}' is set to an existing column within the schema.", n))
+                })
+                .transpose()?;
+        if let Some(index_column) = index_column {
+            properties.insert(ES_OPTION_INDEX_COLUMN.to_string(), index_column.to_string());
+        }
+        let config = serde_json::from_value::<ElasticSearchOpenSearchConfig>(
+            serde_json::to_value(properties).unwrap(),
+        )
+        .map_err(|e| SinkError::Config(anyhow!(e)))?;
+        Ok(config)
+    }
+
+    pub fn build_elasticsearch_client(&self) -> Result<elasticsearch::Elasticsearch> {
+        let url =
+            Url::parse(&self.url).map_err(|e| SinkError::ElasticSearchOpenSearch(anyhow!(e)))?;
+        let transport = elasticsearch::http::transport::TransportBuilder::new(
+            elasticsearch::http::transport::SingleNodeConnectionPool::new(url),
+        )
+        .auth(elasticsearch::auth::Credentials::Basic(
+            self.username.clone(),
+            self.password.clone(),
+        ))
+        .build()
+        .map_err(|e| SinkError::ElasticSearchOpenSearch(anyhow!(e)))?;
+        let client = elasticsearch::Elasticsearch::new(transport);
+        Ok(client)
+    }
+
+    pub fn build_opensearch_client(&self) -> Result<opensearch::OpenSearch> {
+        let url =
+            Url::parse(&self.url).map_err(|e| SinkError::ElasticSearchOpenSearch(anyhow!(e)))?;
+        let transport = opensearch::http::transport::TransportBuilder::new(
+            opensearch::http::transport::SingleNodeConnectionPool::new(url),
+        )
+        .auth(opensearch::auth::Credentials::Basic(
+            self.username.clone(),
+            self.password.clone(),
+        ))
+        .build()
+        .map_err(|e| SinkError::ElasticSearchOpenSearch(anyhow!(e)))?;
+        let client = opensearch::OpenSearch::new(transport);
+        Ok(client)
+    }
+}
+
+pub struct ElasticSearchOpenSearchFormatter {
+    key_encoder: TemplateEncoder,
+    value_encoder: JsonEncoder,
+    index_column: Option<usize>,
+    index: Option<String>,
+}
+
+type BuildBulkPara = (String, String, Option<Map<String, Value>>);
+
+impl ElasticSearchOpenSearchFormatter {
```

Comment:
For the original java implementation, we already had rust code that convert the stream chunk into formatted json value. Can we reuse the code rather than implementing the logic again?